components/server/src/websocket/websocket-connection-manager.ts

```diff
@@ -83,19 +137,23 @@ export class WebsocketConnectionManager<C extends GitpodClient, S extends Gitpod
             userAgent: expressReq.headers["user-agent"],
             dnt: takeFirst(expressReq.headers.dnt),
             clientRegion: takeFirst(expressReq.headers["x-glb-client-region"]),
-        }
+        };
 
         gitpodServer.initialize(client, user, resourceGuard, clientHeaderFields);
         client.onDidCloseConnection(() => {
-            increaseApiConnectionClosedCounter();
             gitpodServer.dispose();
+            increaseApiConnectionClosedCounter();
+            this.events.emit(EVENT_CONNECTION_CLOSED, gitpodServer, expressReq);
 
-            this.removeServer(gitpodServer);
-            this.events.emit(EVENT_CONNECTION_CLOSED, gitpodServer);
+            clientContext.removeEndpoint(gitpodServer);
+            if (clientContext.hasNoEndpointsLeft()) {
+                this.contexts.delete(clientContext.clientId);
+                this.events.emit(EVENT_CLIENT_CONTEXT_CLOSED, clientContext);
+            }
         });
-        this.servers.push(gitpodServer);
+        clientContext.addEndpoint(gitpodServer);
```

Comment:
this still initializes a new server per client, right?

Reply:
Exactly. It's still connection <-> server.

The `WebsocketUserContext` does not surface a purpose atm, despiste making the connection visible in code.
Paths we could extend this are:
 - making cached data visible (currently weaved in implicitly using inversify using `inRequestScope` :roll_eyes: )
 - for additional reporting (a histogram with connections/user)
 - manage/distribute user-level messages